torch/csrc/distributed/c10d/CUDASymmetricMemoryOps.cu

```diff
@@ -693,6 +695,123 @@ at::Tensor stream_write_value32_(
   return input;
 }
 
+std::tuple<at::Tensor, at::Tensor> _all_gather_mm(
+    at::Tensor a_shard,
+    at::Tensor b,
+    std::string group_name) {
+  c10::cuda::CUDAGuard guard(a_shard.device());
+
+  auto symm_mem = c10d::symmetric_memory::rendezvous(a_shard, group_name);
+  TORCH_CHECK(
+      symm_mem != nullptr,
+      "_all_gather_mm: input must be allocated with empty_strided_p2p().");
+
+  int rank = symm_mem->get_rank();
+  int world_size = symm_mem->get_world_size();
+
+  auto current_stream = at::cuda::getCurrentCUDAStream();
+
+  static std::optional<at::cuda::CUDAStream> copy_stream = std::nullopt;
+  static std::optional<at::cuda::CUDAStream> signal_stream = std::nullopt;
+  if (!copy_stream.has_value()) {
+    copy_stream = c10::cuda::getStreamFromPool(-1, a_shard.device().index());
+    signal_stream = c10::cuda::getStreamFromPool(-1, a_shard.device().index());
+  }
+
+  symm_mem->barrier(0, 0);
+  {
+    auto event = at::cuda::CUDAEvent();
+    event.record(current_stream);
+    event.block(*copy_stream);
+  }
+
+  auto is_capturing =
+      (at::cuda::currentStreamCaptureStatus() != at::cuda::CaptureStatus::None);
+  if (is_capturing) {
+    auto event = at::cuda::CUDAEvent();
+    event.record(current_stream);
+    event.block(*signal_stream);
+  }
+
+  auto a_sizes = a_shard.sizes().vec();
+  a_sizes[0] *= world_size;
+  auto a = a_shard.new_empty(a_sizes);
+  auto a_signals = at::zeros(
+      world_size,
+      at::TensorOptions().dtype(at::ScalarType::UInt32).device(a.device()));
+
+  auto shard_sz = a_shard.numel() * a_shard.element_size();
+  std::vector<void*> chunk_ptrs;
+  chunk_ptrs.reserve(world_size);
+  for (int r = 0; r < world_size; ++r) {
+    chunk_ptrs.push_back(
+        reinterpret_cast<uint8_t*>(a.data_ptr()) + r * shard_sz);
+  }
+
+  // stream_write_value32_ is more efficient than memset32_, but it is not
+  // compatible with CUDA graph. Using memset32_ directly on copy_stream during
```

Comment:
I find this surprising, as the CUDA driver has a `cuGraphAddBatchMemOpNode` function precisely to add such nodes to the graph. Perhaps this is only supported by capturing if one uses the `cuStreamBatchMemOp` function (to add a group of ops) instead of adding them individually via `cuStreamWriteValue32`?